### ISSUE TITLE
navigation.state.params is NULL when navigating out from parent route

### ISSUE THREAD

**Issue by ChrisEelmaa** [react-navigation/react-navigation#6785]
Note I wasn't sure to which place to make the issue to. It could be related to tab navigator, however I believe this to be more general issue/limitation of react-navigation library.

I am using v5 compat mode, although different versions of library have different behaviour. Why I have set up the following structure in my app is simple, that is the only structure that has allowed to me to have "modals" (with proper animation as well) inside tab navigator.

```


const AppTabsNavigator = createCompatNavigatorFactory(createStackNavigator)({

    /* Just a tab structure */
    MyMatches: createCompatNavigatorFactory(createMaterialTopTabNavigator)({
        MyMatches: createCompatNavigatorFactory(createStackNavigator)({
          MyMatches: MyMatchesScreen,
          MyMatchChat: MyMatchChatScreen,
        }, {
              initialRouteName: 'MyMatches'
          });,
    },

    {
        initialRouteName: 'MyMatches',

        tabBar: () => null,
        swipeEnabled: false,
    }),

   /* Just a react component, */
    ReportAndUnmatchWhatDoYouWantToDoModal: ReportAndUnmatchWhatDoYouWantToDoModal

}
{
    initialRouteName: 'MeetPeople',
});

```

Looking from top-down:

-- StackNavigator (headermode: 'none, mode: 'modal')
---TabNavigator
---- My Matches (Stack Navigator)
----- My matches chat (react component)

--- ReportAndUnmatchWhatDoYouWantToDoModal (react component)

What is the problem? 

*One problem* is when I navigate from 'My matches chat' to ReportAndUnmatchWhatDoYouWantToDoModal and pass parameters, ReportAndUnmatchWhatDoYouWantToDoModal component does not receive any parameters at all, I have tried every possible method / property there exist, including:

```
    alert(JSON.stringify(this.props.navigation, null, 4) + "along" +JSON.stringify(this.props.route, null, 4)  +
    JSON.stringify(this.props.navigation.state, null, 4)
```

I've also tried dangerouslyGetParent, along with dangerouslyGetState(), they all report no parameters, essentially parameters are not getting passed anymore. This used to work in v3 and v4, but stopped working in v5 compat.

Is there any advice that can be given? Cheers

**Comment by ChrisEelmaa** [comment]
here is the repro:

``` 
git clone https://github.com/ChrisEelmaa/react-navigaiton-tabnavigator-repro
npm install
npm run start (to launch live reload server)
npm run ios (in another terminal)
```

when the flow gets to `ReportAndUnmatchWhatDoYouWantToDoModal`, the parameter is not passed. (`WHAT: 10`), this used to work in previous versions and I could pass the parameters with the same structure.

Also getting warning in the console: 

```
 "Accessing child navigation state for a route is not safe and won't work correctly."
```

**Comment by Joaojuby** [comment]
I've run into this issue today while migrating from v4 to v5.
I was using the compat tools and faced the exact same issue where passing a parameter to a tab navigator was not passing it down to the tabs themselves **unless** I specifically navigate to that tab and assign it the params.

I understand from the documentation that this is actually the intended behaviour but it caused me and my team some issues. As I see it the pattern is that I need that parameter in all the tabs but the library stopped passing down after the upgrade to v5.

To change this we had a few options:
 - Setting the parameter in a redux store and read from it (navigation agnostic but requires us to re-do all navigation actions, refactor the screens to fetch from the store and extensively test navigation actions and deep links handling (we don't handle deep links with react-navigation)
 - Change the actions for nested navigators using the new API, block swipes (poor UX choice for us) and change the tabs press actions to also navigate with the params for the respective tab screens (we also have customized tab bars so this might not be an option for others)
 - Migrate the needed stacks directly to v5 hoping it would solve the issue

We actually chose the last one and while that wasn't the solution to our problems, we found a solution.
First it wasn't a solution because the parent navigator doesn't pass the params to the children as in v4. Sure that's not great, but thanks to the v5 API we can actually handle it quite gracefully

Imagine this simple navigation structure
```const AppStackNavigator = createStackNavigator();

const AppStack = () => (
  <AppStackNavigator.Navigator>
    <AppStackNavigator.Screen
      options={{ headerShown: false }}
      name="Root"
      component={MainTabs}
    />
    <AppStackNavigator.Screen
      name="CompetitionsDetails"
      component={CompetitionDetailsTabs}
    />
  </AppStackNavigator.Navigator>
);

export default AppStack;
```

and here's the competitionDetailsTabs code

```const CompetitionDetailsNavigator = createMaterialTopTabNavigator();

const CompetitionDetailsTabs = props => {
  const { competitionId } = props.route.params;
  return (
    <CompetitionDetailsNavigator.Navigator>
      <CompetitionDetailsNavigator.Screen
        initialParams={{ competitionId }}
        name="Standings"
        component={CompetitionStandings}
      />
    </CompetitionDetailsNavigator.Navigator>
  );
};

export default CompetitionDetailsTabs;

```

So the current situation is that we are inside the MainTabs and intend to navigate to the competitions tabs while maintaining all of our legacy actions untouched like this `    this.props.navigation.navigate('CompetitionsDetails', { competitionId });
`

Sorry for the long wall of text but the whole "magic" right now is to grab the props in the navigator, handle it as you see fit and pass it down, it's a developer responsability. We are doing it by assigning the initialParams but we are also hoping it won't cause issues in situations where we "navigate to the same screen" but change the parameters.

Hoping this helps as I had seen a few threads, this one specifically with the same issue and no replies for a few days.
